### PR TITLE
Update LLVM submodule

### LINKED ISSUE #105626: Casting the same i64 and u64 integers to f32 gives different results on `i686-pc-windows-msvc` target
### I tried this code:

```rust
fn main() {
    let mut n: i64 = 0b0000000000111111111111111111111111011111111111111111111111111111;
    println!("{}, {}, {}, {}", n, n as i64 as f32, n as u64 as f32, n as u128 as f32);
    assert_ne!((n as f64) as f32, n as f32);
    assert_eq!(n as i64 as f32, n as u64 as f32);
    n = -n;
    assert_ne!((n as f64) as f32, n as f32);
    assert_eq!(n as i64 as f32, -(-n as u64 as f32));
}
```

rust-num/num-bigint#261

### I expected to see this happen:
The code prints the output below and exits successfully.
```
18014397972611071, 18014397000000000, 18014397000000000, 18014397000000000
```

### Instead, this happened:
The code prints the output below and panics.
```
18014397972611071, 18014397000000000, 18014399000000000, 18014397000000000
```

### Meta

`rustc --version --verbose`:
```
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-pc-windows-msvc
release: 1.65.0
LLVM version: 15.0.0
```

### PR DESCRIPTION
**Pull request by ghost**
Fixes #105626

### CODE REVIEW

**Comment by cuviper** [comment]
Can you add a regression test for this? Probably a ui `run-pass` test with `min-system-llvm-version: 16`.

**Comment by icedrocket** [comment]
I found a problem with number casting tests. If we simply define a variable to check casting with the `assert!`, it will succeed regardless of the LLVM version. However, if we also pass that variable as an argument to the `println!` to call an external function, it fails. `saturating-float-casts-impl.rs` already contains suitable cases to verify that u64 to f32 conversion is correct.

**Comment by icedrocket** [comment]
I found a problem with number casting tests. If we simply define a variable to check casting with the assert!, it will succeed regardless of the LLVM version. However, if we also pass that variable as an argument to the println! to call an external function, it fails. saturating-float-casts-impl.rs already contains suitable cases to verify that u64 to f32 conversion is correct.

**Comment by icedrocket** [comment]
I found a problem with number casting tests. If we simply define a variable to check casting with the `assert!`, it will succeed regardless of the LLVM version. However, if we also pass that variable as an argument to the `println!` to call an external function, it fails. `saturating-float-casts-impl.rs` already contains suitable cases to verify that u64 to f32 conversion is correct.

**Comment by icedrocket** [comment]
https://github.com/rust-lang/rust/blob/eebdfb55fce148676c24555505aebf648123b2de/tests/ui/numbers-arithmetic/saturating-float-casts-impl.rs#L382-L400

It seems that this part doesn't check unsigned to float casting for edge cases. We need to add unsigned to float casting checks and modify `assert_eq` to check both static and runtime casting.

Can I just create another PR to fix the test?

**Comment by cuviper** [comment]
> It seems that this part doesn't check unsigned to float casting for edge cases. We need to add unsigned to float casting checks and modify `assert_eq` to check both static and runtime casting.

I think that's fine to add separately, now that there's a targeted test here.

@bors r+

**Comment by cuviper** [comment]
Those values failed equality on `i586-unknown-linux-gnu`, but print identically -- that might be more x87 excess precision at work on i586. Maybe we should just limit this test to `only-i686-pc-windows-msvc`?

**Comment by icedrocket** [comment]
I force-pushed to parent commit by accident.

**Comment by icedrocket** [comment]
I force-pushed to parent commit by accident. Now ready for review.

**Comment by icedrocket** [comment]
I force-pushed to parent commit by accident.

**Comment by icedrocket** [comment]
I force-pushed to parent commit by accident. Now ready for review.

**Comment by cuviper** [comment]
I don't think `black_box` is a reliable way to avoid x87 issues. What was wrong with your `sse2` version of the test? This regression test is specifically for a particular target platform; it's ok if we don't run that test elsewhere.

**Comment by icedrocket** [comment]
> I don't think `black_box` is a reliable way to avoid x87 issues. What was wrong with your `sse2` version of the test? This regression test is specifically for a particular target platform; it's ok if we don't run that test elsewhere.

The u64 to f32 conversion is currently implemented using x87 even when SSE2 is enabled. If in the future the compiler changes to use SSE2 for that conversion, it will be difficult to know if there is a problem with the x87 implementation as the test will not fail as long as SSE2 is enabled.

**Comment by icedrocket** [comment]
I will create a separate PR to remove the SSE2 flag after the x87 issues resolved.

**Comment by icedrocket** [comment]
I'll create a separate PR to remove the `sse2` feature flag after the x87 issues resolved.

**Comment by icedrocket** [comment]
I will create a separate PR to remove the `sse2` feature flag after the x87 issues resolved.

**Comment by icedrocket** [comment]
But I think it's okay for us to just use the SSE2 version for now. As long as SSE2 is enabled, we can generate code that works fine on x86. I will contribute to solving the x87 issue on the LLVM side later.

**Comment by cuviper** [comment]
The LLVM fix is specific to Windows, and `i686-pc-windows-msvc` does have SSE2 enabled by default. I'm not sure what else you're hoping to fix in LLVM, but I think we're good here.

@bors r+

**Comment by cuviper** [comment]
[`(x86_64-gnu-llvm-14, 1, ubuntu-20.04-xl)`](https://github.com/rust-lang-ci/rust/actions/runs/4292689140/jobs/7479447058#step:26:36811):
> ```
> ---- config::tests::download_ci_llvm stdout ----
> Detected LLVM as non-available: running in CI and modified LLVM in this change
> Detected LLVM as non-available: running in CI and modified LLVM in this change
> Detected LLVM as non-available: running in CI and modified LLVM in this change
> Detected LLVM as non-available: running in CI and modified LLVM in this change
> thread 'config::tests::download_ci_llvm' panicked at 'assertion failed: parse_llvm(\"build.build = \\\"x86_64-unknown-linux-gnu\\\"\")', config/tests.rs:22:5
> error: test failed, to rerun pass `--lib`
> ```

@rust-lang/bootstrap -- this should be expected! We're updating LLVM, but not all CI builders use that.

**Comment by jyn514** [comment]
@cuviper oops, this might be the first time LLVM's been updated since https://github.com/rust-lang/rust/pull/107234 - I think we should skip this test altogether if LLVM was updated in the PR. Maybe take the code in https://github.com/rust-lang/rust/blob/ae5473c9699eaf1998c414e0a8884383bb9ddcc3/src/bootstrap/native.rs#L219-L230 and separate it into a function?

**Comment by cuviper** [comment]
@jyn514 I gave that a shot, let me know what you think!

File: src/bootstrap/native.rs [hunk 1]
```diff
@@ -216,21 +216,26 @@ pub(crate) fn is_ci_llvm_available(config: &Config, asserts: bool) -> bool {
...
 
-    if CiEnv::is_ci() {
+    if is_ci_llvm_modified(config) {
+        eprintln!(
+            "Detected LLVM as non-available: running in CI and modified LLVM in this change"
+        );
+        return false;
+    }
+
+    true
```
Comment (albertlarsan68) [inline src/bootstrap/native.rs:226]: Definitly non-blocking, just personal hate with `return`.

```suggestion
    if is_ci_llvm_modified(config) {
        eprintln!(
            "Detected LLVM as non-available: running in CI and modified LLVM in this change"
        );
        false
    } else {
        true
    }
```

**Review by albertlarsan68 — APPROVED** [review]
I approve the last commit (91af58ac58dee5794fe54a8f9cb5556d3585688d).
Just a minor nit, if you want.

File: src/bootstrap/native.rs (hunk 1 above)
Comment (cuviper) [reply src/bootstrap/native.rs]: I can change that if you want, but note that this is one block in a (short) series of early returns. That's a pattern where I think it makes sense not to directly chain `if-else`, especially since the first one is a more complex nested return.

**Review by jyn514 — APPROVED** [review]

File: src/bootstrap/native.rs (hunk 1 above)
Comment (ozkanonur) [reply src/bootstrap/native.rs]: I would avoid nested code blocks as much as possible. I don't think having `else` helping the code look better and easier to read

Comment (cuviper) [reply src/bootstrap/native.rs]: I've left it alone for now.
